test/core/client_channel/lb_policy/round_robin_test.cc

```diff
@@ -0,0 +1,373 @@
+  ExpectReresolutionRequest();
+  ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectNoStateChange();
+}
+
+TEST_F(RoundRobinTest, AllTransientFailure) {
+  std::array<std::string, 3> uris = {
+      "ipv4:127.0.0.1:441",
+      "ipv4:127.0.0.1:442",
+      "ipv4:127.0.0.1:443",
+  };
+  auto policy = MakeLbPolicy("round_robin");
+  auto status = ApplyUpdate(BuildUpdateArgs(uris), policy.get());
+  ASSERT_TRUE(status.ok()) << status;
+  for (int i = 0; i < 3; i++) {
+    ExpectState(GRPC_CHANNEL_CONNECTING);
+  }
+  ExpectState(GRPC_CHANNEL_CONNECTING);
+  for (size_t i = 0; i < uris.size() - i; ++i) {
+    auto subchannel = FindSubchannel(uris[i]);
+    ASSERT_NE(subchannel, nullptr);
+    subchannel->SetConnectivityState(GRPC_CHANNEL_TRANSIENT_FAILURE,
+                                     absl::OkStatus());
+    ExpectReresolutionRequest();
+    ExpectState(GRPC_CHANNEL_CONNECTING);
+  }
+  FindSubchannel(uris[2])->SetConnectivityState(GRPC_CHANNEL_TRANSIENT_FAILURE,
+                                                absl::OkStatus());
+  ExpectReresolutionRequest();
+  ExpectState(GRPC_CHANNEL_TRANSIENT_FAILURE);
+  ExpectNoStateChange();
+}
+
+TEST_F(RoundRobinTest, NoChannels) {
+  auto policy = MakeLbPolicy("round_robin");
+  LoadBalancingPolicy::UpdateArgs update_args;
+  update_args.resolution_note = "This is a test";
+  update_args.addresses.emplace();
+  absl::Status status = ApplyUpdate(std::move(update_args), policy.get());
+  EXPECT_TRUE(absl::IsUnavailable(status));
+  ExpectState(GRPC_CHANNEL_TRANSIENT_FAILURE,
+              absl::UnavailableError("empty address list: This is a test"));
+}
+
+TEST_F(RoundRobinTest, AddressListChange) {
+  std::array<std::string, 3> uris = {
+      "ipv4:127.0.0.1:441",
+      "ipv4:127.0.0.1:442",
+      "ipv4:127.0.0.1:443",
+  };
+  auto policy = MakeLbPolicy("round_robin");
+
+  // One channel that is ready. Gets picked all the time.
+  auto status = ApplyUpdate(BuildUpdateArgs({uris[0]}), policy.get());
+  ASSERT_TRUE(status.ok()) << status;
+  ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectNoStateChange();
+  FindSubchannel(uris[0], {})
+      ->SetConnectivityState(GRPC_CHANNEL_READY, absl::OkStatus());
+  auto picker = ExpectState(GRPC_CHANNEL_READY);
+  auto picked = ExpectPickAddress(picker.get());
+  EXPECT_EQ(*picked, uris[0]);
+  ExpectNoStateChange();
+
+  // Second channel added, connecting. Only the first channel gets picked still
+  status = ApplyUpdate(BuildUpdateArgs({uris[0], uris[1]}), policy.get());
+  ASSERT_TRUE(status.ok()) << status;
+  ExpectState(GRPC_CHANNEL_READY);
+  picker = ExpectState(GRPC_CHANNEL_READY);
+  EXPECT_EQ(*ExpectPickAddress(picker.get()), uris[0]);
+  EXPECT_EQ(*ExpectPickAddress(picker.get()), uris[0]);
+  ExpectNoStateChange();
+
+  // Second channel ready. Both channels are now picked.
+  FindSubchannel(uris[1], {})
+      ->SetConnectivityState(GRPC_CHANNEL_READY, absl::OkStatus());
```

Comment:
This new subchannel is not going to go from IDLE (the state it reports when it is first created) directly to READY.  We should either (a) use `CreateSubchannel()` to add the subchannel in the subchannel pool and set its state before we send the update containing the new address to the LB policy, or (b) go through the normal sequence of having the LB policy request a connection on the subchannel and then transitioning first to CONNECTING before then moving to READY.

Reply:
done